### PR TITLE
Reduce minimum icon size to "smallest" in List View

### LINKED ISSUE #1164: Too large icons in List-view
Icons are to too big by default. As such I feel that the 'list-view' becomes less effective.  Using dconf-editor I changed all zoom-levels to 'smallest'. Much better!

### PR DESCRIPTION
**Pull request by jeremypw**
Fixes #1164
There does not seem to be a good reason to maintain this restriction on the user.

In order to gain the benefits in compactness possible with a smaller icon size, the line spacing has been reduced.  At all but the smallest icon size the icon height dominates in anycase. 

In addition, the layout of the icon, helper and emblems has been tweaked so that less of the icon is obscured and enabling showing of the emblems at smallest icon size.

### CODE REVIEW

**Comment by danrabbit** [comment]
It looks like at the smallest zoom in list view we lose icon badges such as the symlink or read-only emblems.

It's also weird that this is a dconf setting, so changing this won't actually fix the issue for users with existing installs since we don't overwrite user preferences on update

**Comment by jeremypw** [comment]
How about removing the setting altogether for the List and Column Views?  There is a different PR outstanding that tries to address a similar issue with Icon View.

**Comment by jeremypw** [comment]
Regarding the emblems, this was done to avoid completely obscuring the icon with the overlapped emblems.  In the linked issue I suggested maybe the emblems should be moved to the side (non-overlapping) at the smallest icon size instead of hiding but I did not want to complicate this PR.

**Comment by danrabbit** [comment]
Yeah I agree we should remove the setting. It doesn't really seem like there's a good reason for this to be configurable. We should support the sizes we support :)

My feeling is that we should probably address issues with the smallest icon size before making it generally available for people to use

**Comment by jeremypw** [comment]
I am not sure the hiding is really a problem (or not a new one) as this has happened in Column View for a long time with no issues raised.

**Comment by jeremypw** [comment]
I have now removed the "minimum-zoom" setting altogether for the two tree views.  I have left for IconView as this does not cope well with the smallest icon sizes.  There is another PR outstanding that attempts to produce a usable IconView at small icon sizes - see #1318

**Comment by jeremypw** [comment]
AFTER 
Screenshot from 2020-07-18 19 07 50

**Comment by jeremypw** [comment]
BEFORE
Screenshot from 2020-07-18 19 13 47

**Comment by jeremypw** [comment]
@danrabbit I have attempted to improve the situation with emblems and small icons, as you suggested.  I have increased the xpadding and used the extra space to reduce the amount that the emblems overlap the icon at small icon sizes.

File: src/View/AbstractTreeView.vala [hunk 1]
```diff
@@ -349,6 +352,49 @@ namespace FM {
...
+                        break;
+                }
+
+                try {
+                    view_css.load_from_data ("treeview { -GtkTreeView-vertical-separator: %i px; }".printf (vsep));
+                    get_style_context ().add_provider (view_css, Gtk.STYLE_PROVIDER_PRIORITY_APPLICATION);
+                } catch (Error e) {
+                    warning ("Could not style FM.TreeView");
+                }
+            }
```
Comment (danrabbit) [inline src/View/AbstractTreeView.vala:385]: I would really prefer not to do this. If the padding is too large for tree view rows, we should address that in the stylesheet and not special case Files styling

File: src/View/ColumnView.vala [hunk 2]
```diff
@@ -56,21 +56,20 @@ namespace FM {
             return false;
         }
 
-        protected override Marlin.ZoomLevel get_set_up_zoom_level () {
-            var zoom = Marlin.column_view_settings.get_enum ("zoom-level");
-            Marlin.column_view_settings.bind ("zoom-level", this, "zoom-level", GLib.SettingsBindFlags.SET);
+        protected override void set_up_zoom_level () {
+            Marlin.column_view_settings.bind ("zoom-level",
+                                                           this, "zoom-level",
+                                                           GLib.SettingsBindFlags.DEFAULT);
```
Comment (danrabbit) [inline src/View/ColumnView.vala:62]: I don't really think this is easier to read than having it inline, but if it must be shortened then code style is for it to be done like:
```suggestion
            Marlin.column_view_settings.bind (
                "zoom-level",
                this, "zoom-level",
                GLib.SettingsBindFlags.DEFAULT
            );
```

File: src/View/IconView.vala [hunk 3]
```diff
@@ -86,9 +87,10 @@ namespace FM {
             return tree as Gtk.Widget;
         }
 
-        protected override Marlin.ZoomLevel get_set_up_zoom_level () {
-            var zoom = Marlin.icon_view_settings.get_enum ("zoom-level");
-            Marlin.icon_view_settings.bind ("zoom-level", this, "zoom-level", GLib.SettingsBindFlags.SET);
+        protected override void set_up_zoom_level () {
+            Marlin.icon_view_settings.bind ("zoom-level",
```
Comment (danrabbit) [inline src/View/IconView.vala:91]: same as above for code style

File: src/View/ListView.vala [hunk 4]
```diff
@@ -198,22 +198,20 @@ namespace FM {
             return tree as Gtk.Widget;
         }
 
-        protected override Marlin.ZoomLevel get_set_up_zoom_level () {
-            var zoom = Marlin.list_view_settings.get_enum ("zoom-level");
-            Marlin.list_view_settings.bind ("zoom-level", this, "zoom-level", GLib.SettingsBindFlags.SET);
+        protected override void set_up_zoom_level () {
+            Marlin.list_view_settings.bind ("zoom-level",
```
Comment (danrabbit) [inline src/View/ListView.vala:202]: same code style comment as above

**Review by danrabbit — DISMISSED** [review]
Left a few inline comments

I'm not a fan of injecting custom CSS in this way. If the padding for treeview rows is too big, we should address that at the stylesheet level and not make Files inconsistent with other apps

File: src/View/AbstractTreeView.vala (hunk 1 above)
Comment (jeremypw) [reply src/View/AbstractTreeView.vala:385]: I agree in principle - this is more by way of a demonstration.  The padding only changes for small icons - can you accomplish that in the stylesheet? Would you add two new classes for the treeview with different padding?

File: src/View/IconView.vala (hunk 3 above)
Comment (jeremypw) [reply src/View/IconView.vala:91]: Theres a lot of old code where this style needs updating!

File: src/View/AbstractTreeView.vala (hunk 1 above)
Comment (danrabbit) [reply src/View/AbstractTreeView.vala:385]: I'm not sure we really need to have two different paddings. Maybe we can open an issue in the stylesheet to reconsider the default and see where we wind up?

Comment (jeremypw) [reply src/View/AbstractTreeView.vala:385]: The row padding is reduced at small icon size to get the maximum benefit in terms of compactness a well as avoiding an unbalanced appearance (space between rows too large compared with the row content).  There are in fact three different paddings used in this PR (0, 3 and 5 px).  But I agree it is ultimately up to the UX team to arrive at the appropriate compromise. I'll open an issue in stylesheet.  Is this PR mergeable if I remove the CSS or shall I put it on the backburner?

**Comment by jeremypw** [comment]
Smallest icon size without CSS to reduce row padding:
Screenshot from 2020-08-20 10-23-40

**Comment by jeremypw** [comment]
@danrabbit - Is it worth pursuing this?

**Comment by danrabbit** [comment]
Hey sorry I haven't followed up here in a while. I thought I just had was maybe in the list and column views we should treat emblems as their own columns instead of badging the icon:

badges

**Comment by jeremypw** [comment]
A problem with putting the emblems in their own column is that the emblems would be shown  vertically aligned outside the right edge of the filename column in the view therefore potentially far from a short name.   It would also result in up to 4 column headers for usually empty columns (the current implementation caters for up to 4 emblems).

It may be possible to change the TextRenderer class to render the emblems just after the filename, or have the IconRenderer show them differently in the icon column, but I think that would be out of scope for this PR and should be the subject of a new issue.

**Comment by jeremypw** [comment]
@danrabbit So, how do you want to proceed with this PR?

**Review by danrabbit — APPROVED** [review]
Does what it intends to do and ditches some unecessary gsettings stuff. We can file other issues and follow up on making list views better in other branches I think :)

**Pull request merged**